core/src/main/java/de/danoeh/antennapod/core/service/download/DownloadService.java

```diff
@@ -178,6 +180,8 @@ public int onStartCommand(Intent intent, int flags, int startId) {
     @Override
     public void onCreate() {
         Log.d(TAG, "Service started");
+        newEpisodesNotification = new NewEpisodesNotification(getApplicationContext());
```

Comment:
Sorry for all the back and forth. I just noticed a problem with this method. This executes database operations on the main thread, which could lead to problems when having a big number of feeds. How about moving notification handling to `handleSuccessfulDownload`? That one is executed in the background. Also, the `FeedSyncTask` that it called there needs to load the feed anyway. Maybe that task could get another property that shows if there was a new item.

Reply:
Not sure if I understood this correctly... like this?

Reply:
The problem is that you call the `NewEpisodesNotification` constructor in `onCreate`. This causes database operations like `DBReader.getFeedItemList` to be executed on the main thread (which can cause "application not responding" errors if it takes too long). Currently, you handle all subscriptions at once. My suggestion would be to handle the subscriptions one by one in `FeedSyncTask` (called from `handleSuccessfulDownload`). That way, loading the items is done in a background thread (because `handleSuccessfulDownload` is called in a background thread). Doing this in `FeedSyncTask` has another advantage: The class loads the feed from the database anyway. So when using that class, AntennaPod needs to do fewer database queries, which is good for the performance.

Reply:
This took me some time because there are some issues, which seam hard to fix to me:
- because this is split up now, there is no unified place to create the group summary from. also I would have to create another data class for holding the Group summary and pass it around like a hot potato between the classes and methods. (thats why the group summary has currently been removed)
- because this is split up and slightly asynchronous I fear it might have more than one notification sound now.
- forgot.. to be added

Reply:
I will have a look at the code changes later but I already have some comments to your message :)

> because this is split up and slightly asynchronous I fear it might have more than one notification sound now.

Could something like this help? I have not tested it but it looks like it is possible to only have one notification sound per group. https://developer.android.com/reference/androidx/core/app/NotificationCompat.Builder#setGroupAlertBehavior(int)

> because this is split up now, there is no unified place to create the group summary from

Maybe something like "You have new episodes" without the number could be enough (and is easier to implement).

> I can't fully see where I would use some db connection

Anything that calls DbReader or DbWriter should not be called on the UI thread.

Reply:
> Maybe something like "You have new episodes" without the number could be enough (and is easier to implement).

This does not work because:

> To support older versions, which cannot show a nested group of notifications, you must create an extra notification that acts as the summary. This appears as the only notification [...]. So this **summary should include a snippet from all the other notifications**, which the user can tap to open your app.

So in our case users with pre Android 7.0 would only see that some number of podcast have some amount of new episodes. And the only action they have is to open the app. I think In this case I would prefere not having the ability to Group up the notifications and delete them all at once over only having a vague something.

Reply:
Hmm, maybe we could create a single notification manager object in `DownloadService` that does something like this in onCreate?

    syncExecutor.execute(() -> newEpisodeNotification.loadCounters());

`syncExecutor` is a single background thread, so if posting the method call to that thread is the first thing that the `DownloadService` does, we can be sure that the counters are loaded before the first feed is parsed.

Then it could do something like this in `handleSuccessfulDownload`:

    newEpisodeNotification.displayIfNecessary(request);

The problem is that we can not really do database things in onDestroy. The method itself is called on the main thread and it needs to shut down the executors. So the easiest method would be to do this in `handleSuccessfulDownload`, I think.